File: source/javascript/file_defer_close.js
// File defer close - write, read back, verify
const fs = require('fs');

function main() {
    const path = '/tmp/js_defer_test.txt';
    fs.writeFileSync(path, 'hello from defer');
    try {
        const fd = fs.openSync(path, 'r');
        try {
            const contents = fs.readFileSync(fd, 'utf8');
            if (contents === 'hello from defer') {
                console.log('wrote and read back successfully');
            }
        } finally {
            fs.closeSync(fd);
        }
    } finally {
        fs.unlinkSync(path);
    }
}

main();

File: source/c/file_defer_close.c
// gcc -o file_defer_close file_defer_close.c
#include <stdio.h>
#include <string.h>

int main(void) {
    const char *path = "/tmp/hemlock_defer_test.txt";

    // Write file
    FILE *fw = fopen(path, "w");
    if (fw) { fputs("hello from defer", fw); fclose(fw); }

    // Read file (defer f.close() → fclose at end of scope)
    FILE *f = fopen(path, "r");
    if (f) {
        char contents[256] = {0};
        fgets(contents, sizeof(contents), f);
        fclose(f); // defer equivalent
        if (strcmp(contents, "hello from defer") == 0)
            printf("wrote and read back successfully\n");
    }

    remove(path);
    return 0;
}
